## Frequency Distribution

Frequency distributions are visual displays that organise and present frequency counts so that the information can be interpreted more easily. Frequency distributions can show absolute frequencies or relative frequencies, such as proportions or percentages.

In [2]:
import pandas as pd
import matplotlib.pyplot as plt
path = "data/wnba.csv"
wnba = pd.read_csv(path)

To generate a frequency distribution table using Python, we can use the `Series.value_counts()` method. Let's try it on the Pos column, which describes the position on the court of each individual.

In [3]:
wnba['Pos'].value_counts()

G      60
F      33
C      25
G/F    13
F/C    12
Name: Pos, dtype: int64

for height

In [4]:
wnba['Height'].value_counts()

188    20
193    18
175    16
185    15
183    11
173    11
191    11
196     9
178     8
180     7
170     6
198     5
201     2
168     2
206     1
165     1
Name: Height, dtype: int64

## Sorting Frequency distribution tables

Pandas sort tables by default inorder of descending frequencies

This default is harmless for variables measured on a nominal scale because the unique values, although different, have no direction (we can't say, for instance, that centers are greater or lower than guards). The default actually helps because we can immediately see which values have the greatest or lowest frequencies, we can make comparisons easily, etc.

For variables measured on ordinal, interval, or ratio scales, this default makes the analysis of the tables more difficult because the unique values have direction (some unique values are greater or lower than others). Let's consider the table for the Height variable, which is measured on a ratio scale:
Consider the distortation of the Height variables above

Because the Height variable has direction, we might be interested to find:

+ How many players are under 170 cm?
+ How many players are very tall (over 185)?
+ Are there any players below 160 cm?

It's time-consuming to answer these questions using the table above. The solution is to sort the table ourselves.

`wnba['Height'].value_counts()` returns a Series object with the measures of height as indices. This allows us to sort the table by index using the Series.`sort_index()` method:

In [5]:
wnba['Height'].value_counts().sort_index()

165     1
168     2
170     6
173    11
175    16
178     8
180     7
183    11
185    15
188    20
191    11
193    18
196     9
198     5
201     2
206     1
Name: Height, dtype: int64

We can also sort the table by index in descending order using

In [6]:
wnba['Height'].value_counts().sort_index(ascending = False)

206     1
201     2
198     5
196     9
193    18
191    11
188    20
185    15
183    11
180     7
178     8
175    16
173    11
170     6
168     2
165     1
Name: Height, dtype: int64

1. Generate a frequency distribution table for the Age variable, which is measured on a ratio scale, and sort the table by unique values.

+ Sort the table by unique values in an ascending order, and assign the result to a variable named age_ascending.
+ Sort the table by unique values in a descending order, and assign the result to a variable named age_descending.

2. Using the variable inspector, analyze one of the frequency distribution tables and brainstorm questions that might be interesting to answer here. These include:

+ How many players are under 20?
+ How many players are 30 or over?


In [7]:
age_ascending = wnba['Age'].value_counts().sort_index()
age_descending = wnba['Age'].value_counts().sort_index(ascending = False)

## Sorting Tables for Ordinal Variables

`Series.sort_index()` doesn't work for ordinal variables as it arranges them in alphabetical order instead of their inherent order

Generate a frequency distribution table for the transformed PTS_ordinal_scale column.

In [8]:
def make_pts_ordinal(row):
    if row['PTS'] <= 20:
        return 'very few points'
    if (20 < row['PTS'] <=  80):
        return 'few points'
    if (80 < row['PTS'] <=  150):
        return 'many, but below average'
    if (150 < row['PTS'] <= 300):
        return 'average number of points'
    if (300 < row['PTS'] <=  450):
        return 'more than average'
    else:
        return 'much more than average'
    
wnba['PTS_ordinal_scale'] = wnba.apply(make_pts_ordinal, axis=1)

In [9]:
wnba['PTS_ordinal_scale'].value_counts()[['very few points', 'few points', 'many, but below average', 
'average number of points','more than average','much more than average']]

very few points             12
few points                  27
many, but below average     25
average number of points    45
more than average           21
much more than average      13
Name: PTS_ordinal_scale, dtype: int64

In [10]:
wnba['PTS_ordinal_scale'].value_counts()[['very few points', 'few points', 'many, but below average', 
'average number of points','more than average','much more than average']].sort_index(ascending = False)

very few points             12
much more than average      13
more than average           21
many, but below average     25
few points                  27
average number of points    45
Name: PTS_ordinal_scale, dtype: int64

## Proportions and Percentages

long way

In [11]:
wnba['Pos'].value_counts() / len(wnba)

G      0.419580
F      0.230769
C      0.174825
G/F    0.090909
F/C    0.083916
Name: Pos, dtype: float64

It's slightly faster though to use Series.value_counts() with the normalize parameter set to True:


In [12]:
wnba['Pos'].value_counts(normalize = True)

G      0.419580
F      0.230769
C      0.174825
G/F    0.090909
F/C    0.083916
Name: Pos, dtype: float64

To find percentages, we just have to multiply the proportions by 100:


In [13]:
wnba['Pos'].value_counts(normalize = True) * 100

G      41.958042
F      23.076923
C      17.482517
G/F     9.090909
F/C     8.391608
Name: Pos, dtype: float64

## Percentalies and percentale ranks

The percentile rank of a given score is the percentage of scores in its frequency distribution that are less than that score. 